File: leetcode/validateStackSequences.js
/**
 * @param {number[]} pushed
 * @param {number[]} popped
 * @return {boolean}
 */
var validateStackSequences = function (pushed, popped) {
  let i = 0;
  let j = 0;
  let s = [];
  while (i < pushed.length || j < popped.length) {
    const top = peek(s);
    if (j < popped.length && top === popped[j]) {
      s.pop();
      j++;
    } else if (i < pushed.length) {
      s.push(pushed[i]);
      i++;
    } else {
      return false;
    }
  }
  return true;
};

var peek = (s) => {
  return s.length ? s[s.length - 1] : null;
};

validateStackSequences([1, 2, 3, 4, 5], [4, 3, 5, 1, 2]);

File: leetcode/validateStackSequences.py
from typing import List


class Solution:
    def validateStackSequences(
            self, pushed: List[int],
            popped: List[int]) -> bool:
        return validate(pushed, popped)


def validate_greedy(pushed: List[int],
                    popped: List[int]) -> bool:
    n = len(pushed)
    stack = []

    j = 0
    for i in range(n):
        stack.append(pushed[i])
        while j < n and stack and stack[-1] == popped[j]:
            stack.pop()
            j += 1
    return j == n


def validate(pushed: List[int],
             popped: List[int]) -> bool:
    i = 0
    j = 0
    n = len(pushed)
    stack = []
    while i < n:
        if pushed[i] == popped[j]:
            i += 1
            j += 1
        elif stack and stack[-1] == popped[j]:
            stack.pop()
            j += 1
        else:
            stack.append(pushed[i])
            i += 1

    while j < n:
        if stack and stack[-1] == popped[j]:
            stack.pop()
            j += 1
        elif stack:
            return False
    return True


validate([1, 2, 3, 4, 5], [4, 3, 5, 1, 2])
